src/audio/detect_test.c

```diff
@@ -0,0 +1,654 @@
+	if (cd->load_memory) {
+		bs = sizeof(uint32_t);
+
+		if (bs > size) {
+			trace_keyword_error("test_keyword_get_model() error: "
+					    "invalid size %d", bs);
+			return -EINVAL;
+		}
+
+		uint32_t crc = crc32(cd->load_memory,
+				     cd->config.load_memory_size);
+
+		trace_keyword("test_keyword_get_model() crc: 0x%X", crc);
+
+		memcpy(cdata->data->data, &crc, bs);
+		cdata->data->abi = SOF_ABI_VERSION;
+		cdata->data->size = bs;
+	} else {
+		trace_keyword_error("test_keyword_get_model() "
+				    "error: invalid cd->config");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int test_keyword_ctrl_get_bin_data(struct comp_dev *dev,
+					  struct sof_ipc_ctrl_data *cdata,
+					  int size)
+{
+	int ret = 0;
+
+	switch (cdata->data->type) {
+	case SOF_DETECT_TEST_CONFIG:
+		ret = test_keyword_get_config(dev, cdata, size);
+		break;
+	case SOF_DETECT_TEST_MODEL:
+		ret = test_keyword_get_model(dev, cdata, size);
+		break;
+	default:
+		trace_keyword_error("test_keyword_ctrl_get_bin_data() error: "
+				   "unknown binary data type");
+	}
+
+	return ret;
+}
+
+static int test_keyword_ctrl_get_data(struct comp_dev *dev,
+				      struct sof_ipc_ctrl_data *cdata, int size)
+{
+	int ret = 0;
+
+	trace_keyword("test_keyword_ctrl_get_data() size: %d", size);
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		ret = test_keyword_ctrl_get_bin_data(dev, cdata, size);
+		break;
+	default:
+		trace_keyword_error("test_keyword_ctrl_get_data() error: "
+				    "invalid cdata->cmd");
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int test_keyword_cmd(struct comp_dev *dev, int cmd, void *data,
+			    int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+
+	trace_keyword("test_keyword_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		return test_keyword_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return test_keyword_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int test_keyword_trigger(struct comp_dev *dev, int cmd)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_trigger()");
+
+	ret = comp_set_state(dev, cmd);
+	if (ret)
+		return ret;
+
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+	case COMP_TRIGGER_RELEASE:
+		cd->detect_preamble = 0;
+		cd->detected = 0;
+		cd->activation = 0;
+		break;
+	}
+
+	return ret;
+}
+
+/*  process stream data from source buffer */
+static int test_keyword_copy(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+
+	tracev_keyword("test_keyword_copy()");
+
+	/* keyword components will only ever have 1 source */
+	source = list_first_item(&dev->bsource_list,
+				 struct comp_buffer, sink_list);
+
+	/* make sure source component buffer has enough data available for copy
+	 * Also check for XRUNs
+	 */
+	if (source->avail < cd->period_bytes) {
```

Comment:
Let's not limit ourselves to period_bytes. We can consume source->avail and return xrun if source->avail == 0.